flow/connectors/utils/lua.go

```diff
@@ -82,31 +81,95 @@ func DefaultOnRecord(ls *lua.LState) int {
 	return 1
 }
 
-type LPool = *puddle.Pool[*lua.LState]
+type LPoolMessage[T any] struct {
+	f   func(*lua.LState) T
+	ret chan<- T
+}
+type LPool[T any] struct {
+	messages chan LPoolMessage[T]
+	returns  chan<- (<-chan T)
+	wait     <-chan struct{}
+	cons     func() (*lua.LState, error)
+	maxSize  int
+	size     int
+	closed   bool
+}
 
-func LuaPool(cons puddle.Constructor[*lua.LState]) LPool {
-	pool, err := puddle.NewPool(&puddle.Config[*lua.LState]{
-		Constructor: cons,
-		Destructor:  (*lua.LState).Close,
-		MaxSize:     4, // TODO env variable
-	})
-	if err != nil {
-		// only happens when MaxSize < 1
-		panic(err)
+func LuaPool[T any](cons func() (*lua.LState, error), merge func(T)) (*LPool[T], error) {
+	maxSize := 4 // TODO env variable
+	returns := make(chan (<-chan T), maxSize)
+	wait := make(chan struct{})
+	messages := make(chan LPoolMessage[T])
+	go func() {
+		for ret := range returns {
+			for val := range ret {
+				merge(val)
+			}
+		}
+		close(wait)
+	}()
+
+	pool := &LPool[T]{
+		messages: messages,
+		returns:  returns,
+		wait:     wait,
+		cons:     cons,
+		maxSize:  maxSize,
+		size:     0,
+		closed:   false,
+	}
+	if err := pool.Spawn(); err != nil {
+		pool.Close()
+		return nil, err
 	}
-	return pool
+	return pool, nil
 }
 
-func LuaPoolJob(ctx context.Context, pool LPool, wait <-chan struct{}, f func(*lua.LState, <-chan struct{})) (chan struct{}, error) {
-	res, err := pool.Acquire(ctx)
+func (pool *LPool[T]) Spawn() error {
```

Comment:
While it'd be fine to spin up a constant N goroutines, the correct amount will depend on the script, making it better to specify that kind of knob on the mirror instead of environment

Instead Spawn is called when messages aren't being waited on, this way an environment can set parallelism to some good amount for the vcores available & fast scripts will only create as many goroutines as necessary to keep up with record ingestion